### PR TITLE
SONARJAVA-2097 move ProgramPoint to top level

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: java-frontend/src/test/java/org/sonar/java/se/SECheckTest.java [hunk 1]
```diff
@@ -38,8 +35,8 @@
   @Test(timeout = 3000)
   public void flow_from_exit_node_should_not_lead_to_infinite_recursion() throws Exception {
     CFG cfg = CFGTest.buildCFG("void foo(boolean a) { if(a) {foo(true);} foo(false); }");
-    ExplodedGraph.Node node = new ExplodedGraph.Node(new ProgramPoint(cfg.blocks().get(3), 0), mock(ProgramState.class));
-    node.addParent(new ExplodedGraph.Node(new ProgramPoint(cfg.blocks().get(2), 2), mock(ProgramState.class)), null);
+    ExplodedGraph.Node node = new ExplodedGraph.Node(new ProgramPoint(cfg.blocks().get(3)), mock(ProgramState.class));
+    node.addParent(new ExplodedGraph.Node(new ProgramPoint(cfg.blocks().get(2)), mock(ProgramState.class)), null);
```
Comment (m-g-sonar) [inline java-frontend/src/test/java/org/sonar/java/se/SECheckTest.java]: This is not equivalent to what was written before, should be the 3rd element.

File: java-frontend/src/test/java/org/sonar/java/se/ProgramPointTest.java [hunk 2]
```diff
@@ -30,9 +30,9 @@
   public void test_to_string_method() throws Exception {
     // ToString method of program point is used by viewer.
     CFG cfg = CFGTest.buildCFG("void foo() {foo();}");
-    ProgramPoint pp = new ProgramPoint(cfg.blocks().get(0), 0);
+    ProgramPoint pp = new ProgramPoint(cfg.blocks().get(0));
     assertThat(pp.toString()).isEqualTo("B1.0  IDENTIFIER1");
-    pp = new ProgramPoint(cfg.blocks().get(0), 2);
+    pp = pp.next().next();
```
Comment (m-g-sonar) [inline java-frontend/src/test/java/org/sonar/java/se/ProgramPointTest.java]: Not really convinced by this notation. I would prefer to have the constructor still being visible, but annotated with `@VisibleForTesting`.

File: java-frontend/src/main/java/org/sonar/java/se/ProgramPoint.java [hunk 3]
```diff
@@ -27,11 +29,23 @@
...
+  }
+
+  private ProgramPoint(CFG.Block block, int i) {
     this.block = block;
     this.i = i;
   }
 
+  ProgramPoint next() {
+    // pointing one element behind is allowed, it represents terminator of the CFG block
+    Preconditions.checkState(i + 1 <= block.elements().size() + 1,
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/ProgramPoint.java]: use a variable for `block.elements().size()`

**Comment by m-g-sonar** [comment]
All your commits are prefixed with "SONARJAVA-2095" when it should be "SONARJAVA-2097"

**Comment by saberduck** [comment]
Fixed wrong JIRA ticket number

File: java-frontend/src/test/java/org/sonar/java/se/ProgramPointTest.java (hunk 2 above)
Comment (saberduck) [inline java-frontend/src/test/java/org/sonar/java/se/ProgramPointTest.java]: As discussed, there would be very few tests using explicit constructor and I prefer to encapsulate the inner state of the class.

File: java-frontend/src/main/java/org/sonar/java/se/ProgramPoint.java (hunk 3 above)
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/ProgramPoint.java]: I added more detailed comment to this one, including one FIXME

**Pull request merged**